src/realm/exec/realm_trawler.cpp

```diff
@@ -515,71 +516,88 @@ std::vector<Entry> Group::get_allocated_nodes() const
 std::vector<FreeListEntry> Group::get_free_list() const
 {
     std::vector<FreeListEntry> list;
-    unsigned sz = m_free_list_positions.size();
-    REALM_ASSERT(sz == m_free_list_sizes.size());
-    REALM_ASSERT(sz == m_free_list_versions.size());
-    for (unsigned i = 0; i < sz; i++) {
-        int64_t pos = m_free_list_positions.get_val(i);
-        int64_t size = m_free_list_sizes.get_val(i);
-        int64_t version = m_free_list_versions.get_val(i);
-        list.emplace_back(pos, size, version);
-    }
-    consolidate_list(list);
+    if (valid()) {
+        unsigned sz = m_free_list_positions.size();
+        REALM_ASSERT(sz == m_free_list_sizes.size());
+        REALM_ASSERT(sz == m_free_list_versions.size());
+        for (unsigned i = 0; i < sz; i++) {
+            int64_t pos = m_free_list_positions.get_val(i);
+            int64_t size = m_free_list_sizes.get_val(i);
+            int64_t version = m_free_list_versions.get_val(i);
+            list.emplace_back(pos, size, version);
+        }
+        consolidate_list(list);
+    }
     return list;
 }
 
-RealmFile::RealmFile(const std::string& file_path, const char* encryption_key)
+RealmFile::RealmFile(const std::string& file_path, const char* encryption_key, bool alternate_top)
 {
     realm::SlabAlloc::Config config;
     config.encryption_key = encryption_key;
     config.read_only = true;
     config.no_create = true;
     m_top_ref = m_alloc.attach_file(file_path, config);
+    if (alternate_top) {
+        m_top_ref = m_alloc.get_alternate_top_ref();
+    }
     m_start_pos = 24;
     m_group = std::make_unique<Group>(m_alloc, m_top_ref);
     m_file_format_version = m_alloc.get_committed_file_format_version();
     std::cout << "Top ref: 0x" << std::hex << m_top_ref << std::dec << std::endl;
     std::cout << "File format version: " << m_file_format_version << std::endl;
+    std::cout << "File size: " << m_alloc.get_baseline() << std::endl;
```

Comment:
normally we use the "logical file size" from the top array when reporting file size

Reply:
Yes, but this value is the actual size of the file. If the top ref is invalid we cannot get the logical file size. The logical file size is printed out from the group class method.